src/validators/dataclass.rs

```diff
@@ -510,10 +517,30 @@ impl Validator for DataclassValidator {
             return self.validate_init(py, self_instance, input, state);
         }
 
-        // same logic as on models
+        // same logic as on models, see more explicit comment in model.rs
         let class = self.class.bind(py);
-        if let Some(py_input) = input_as_python_instance(input, class) {
-            if self.revalidate.should_revalidate(py_input, class) {
+        let generic_origin_class = self.generic_origin.as_ref().map(|go| go.bind(py));
+
+        let (py_instance_input, force_revalidate): (Option<&Bound<'_, PyAny>>, bool) =
+            match input_as_python_instance(input, class) {
+                Some(x) => (Some(x), false),
+                None => {
+                    // if the model has a generic origin, we allow input data to be instances of the generic origin rather than the class,
+                    // as cases like isinstance(SomeModel[Int], SomeModel[Any]) fail the isinstance check, but are valid, we just have to enforce
+                    // that the data is revalidated, hence we set force_revalidate to true
```

Comment:
I'm not 100% sure what happens — the `force_revalidate` here doesn't apply recursively does it, just to the current level right? (I think that's how it should work for generics anyway, though my suspicion is that is how it always works, so probably not an issue, just confirming.)

Reply:
Right, just the current level.